UNIVERSIDAD PROVINCIAL DEL SUDOESTE

ELEMENTOS DE APRENDIZAJE DE MÁQUINA Y BIG DATA


Integrantes:   Grupo 20
•	Lucia ALONSO
•	Pastor PAREDES
•	Viviana SUAREZ
•	Heber ALONSO


LABORATORIO
Análisis de Datos de Ventas Mensuales 
Informe del análisis de un conjunto de datos que contiene información sobre las ventas mensuales de productos en línea. Utilizaremos herramientas de análisis de datos y visualización sobre el desempeño de ventas.


IMPORT Y RUTAS


In [192]:
# from google.colab import drive
# drive.mount('/content/drive')

In [193]:
import os
import pandas as pd
import numpy as np
import plotly.graph_objects as go

# BUSQUEN LA RUTA DONDE ESTAN LOS CSV - EN DROPBOX BIG DATA

# ruta lucia
#ruta = 'C:/Users/luxam/nubes/Dropbox/Estudio/Upso/GRUPO-TRABAJO/BigData/Laboratorio1/Datos/'

# ruta heber
ruta= 'D:/Dropbox/Estudio/Upso/GRUPO-TRABAJO/BigData/Laboratorio1/Datos/'

# ruta viviana
#ruta= '/content/drive/MyDrive/BIG_DATA/DATASET_MESES/Datos'

# ruta pastor
#ruta='C:/Users/pasto/Dropbox/GRUPO-TRABAJO/BigData/Laboratorio1/Datos'


archivos = os.listdir(ruta)
archivos


dataframes ={}
# df_completo = pd.DataFrame()

for archivo in archivos:
    if archivo.endswith('.csv'):  #Asegurarse de que se consideren archivos csv
      mes = archivo.split('_')[3].split('.')[0]  # Obtener el nombre del mes sin la extensión
      ruta_completa = os.path.join(ruta,archivo) # Ruta de acceso completa
      df = pd.read_csv(ruta_completa) # Cargar el csv en un Dataframe
      df_ventas = df.copy()  # Hago copia

      # Transformo todos los valores no numericos en Nan en las columnas Cantidad Pedida y Precio Unitario
      df_ventas['Cantidad Pedida'] = pd.to_numeric(df_ventas['Cantidad Pedida'], errors='coerce')  #El errors= 'coerce' transforma los valores no numericos a NaN'
      df_ventas['Precio Unitario'] = pd.to_numeric(df_ventas['Precio Unitario'], errors='coerce')

      # Elimino todos los Nan y me quedo solo con los valores numericos
      df_ventas = df_ventas.loc[~df_ventas['Cantidad Pedida'].isna()]
      df_ventas = df_ventas.loc[~df_ventas['Precio Unitario'].isna()]

      df_ventas = df_ventas.dropna()

      # Transformo todos los valores numericos al tipo de dato int
      df_ventas['Cantidad Pedida'] = df_ventas['Cantidad Pedida'].astype(int)
      df_ventas['Precio Unitario'] = df_ventas['Precio Unitario'].astype(float)


      dataframes[mes] = df_ventas

df_ventas

ID de Pedido                Producto  Cantidad Pedida  Precio Unitario  \
0           248151   AA Batteries (4-pack)                4             3.84   
1           248152    USB-C Charging Cable                2            11.95   
2           248153    USB-C Charging Cable                1            11.95   
3           248154        27in FHD Monitor                1           149.99   
4           248155    USB-C Charging Cable                1            11.95   
...            ...                     ...              ...              ...   
11681       259353  AAA Batteries (4-pack)                3             2.99   
11682       259354                  iPhone                1           700.00   
11683       259355                  iPhone                1           700.00   
11684       259356  34in Ultrawide Monitor                1           379.99   
11685       259357    USB-C Charging Cable                1            11.95   

      Fecha de Pedido                       Dirección de Envio  
0      09/17/19 14:44      380 North St, Los Angeles, CA 90001  
1      09/29/19 10:19             511 8th St, Austin, TX 73301  
2      09/16/19 17:48    151 Johnson St, Los Angeles, CA 90001  
3      09/27/19 07:52        355 Hickory St, Seattle, WA 98101  
4      09/01/19 19:03            125 5th St, Atlanta, GA 30301  
...               ...                                      ...  
11681  09/17/19 20:56   840 Highland St, Los Angeles, CA 90001  
11682  09/01/19 16:00  216 Dogwood St, San Francisco, CA 94016  
11683  09/23/19 07:39     220 12th St, San Francisco, CA 94016  
11684  09/19/19 17:30   511 Forest St, San Francisco, CA 94016  
11685  09/30/19 00:18   250 Meadow St, San Francisco, CA 94016  

[11629 rows x 6 columns]

Pasos del Análisis

1. Carga de datos:
- Cargar los datos de ventas de cada archivo CSV.

2. Limpieza de los datos:
- Eliminar datos no numéricos.
- Eliminar filas incompletas

3. Preparación de los datos:
- Ajustar los tipos de datos de cada atributo
- Extraer características importantes para el análisis como meses, horas, ciudades etc.

4. Análisis de datos:
Ejemplo:
- Examinar las horas del día en que las ventas son más frecuentes.
- Determinar el mejor momento para mostrar publicidad.

5. Visualización de datos:
- Utilizar gráficos y visualizaciones para representar los resultados de los análisis.
- Crear gráficos de barras, líneas, mapas, tablas, etc, según corresponda.


1. Comportamiento de las ventas en los distintos meses:
● ¿Cómo variaron las ventas a lo largo de los diferentes meses?
● ¿Hubo algún mes que se destacó en términos de ventas?
● ¿Cuál es el ingreso total generado por mes?

Ingreso Generado por mes

In [194]:
meses = ['Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio', 'Julio', 'Agosto', 'Septiembre', 'Octubre', 'Noviembre', 'Diciembre']
dias = list(range(1, 32))

totales_ventas =[] # Ventas Por mes
cantidad_ventas =[] # Cantidad de productos vendidos por mes

for mes in meses:

  dataframe_mes = dataframes[mes]
  dataframe_mes['Total venta'] = dataframe_mes['Cantidad Pedida'] * dataframe_mes['Precio Unitario'] # Genera una columna con la venta total del item
  
  cantidad_ventas_mes = int(dataframe_mes['Cantidad Pedida'].sum())
  cantidad_ventas.append(cantidad_ventas_mes)

  totales_ventas_mes = int(dataframe_mes['Total venta'].sum())  # suma todas las ventas del mes
  totales_ventas.append(totales_ventas_mes) # agrega a la lista la venta total del mes correspondiente
  
  
cantidad_ventas, totales_ventas

([10868,
  13461,
  17006,
  20568,
  18657,
  15255,
  16069,
  13442,
  13119,
  22705,
  19808,
  28121],
 [1815335,
  2203481,
  2809063,
  3396059,
  3144584,
  2578293,
  2646899,
  2244412,
  2098816,
  3736884,
  3198909,
  4619297])

MONTOS TOTALES VENDIDAS POR MES

In [195]:
fig = go.Figure()

# Agrego al objeto figura un grafico de barra conn los valores extraidos previamente
fig.add_trace(go.Bar(
        x= meses,
        y= cantidad_ventas,
        yaxis="y",
        name="Unidades vendidas",
        marker = dict(color= cantidad_ventas, colorscale = 'RdYlGn',colorbar = dict(title = "Total de ventas")),
        # marker = dict(color= 'Light Blue') ,

        text = cantidad_ventas,
        textposition = 'auto'
))

# Configuro titulos y ejes del grafico
fig.update_layout(
    title="Unidades vendidas por mes",
    xaxis=dict(title="Meses"),
    yaxis=dict(title="Unidades vendidas", side="left")
)

# Muestro el grafico
fig.show()

CANTIADES VENIDAS POR MES

In [196]:
fig = go.Figure()

# Agrego al objeto figura un grafico de barra conn los valores extraidos previamente
fig.add_trace(go.Bar(
        x= meses,
        y= totales_ventas,
        yaxis="y",
        name="Monto",
        marker = dict(color= totales_ventas, colorscale = 'RdYlGn',colorbar = dict(title = "Total de ventas")),
        # marker = dict(color= 'Light Blue') ,

        text = totales_ventas,
        textposition = 'auto'
))

# Configuro titulos y ejes del grafico
fig.update_layout(
    title="Montos de Ventas por mes",
    xaxis=dict(title="Meses"),
    yaxis=dict(title="Monto", side="left")
)

# Muestro el grafico
fig.show()

GRAFICO EN TORTA %

In [197]:
import plotly.graph_objects as go
fig = go.Figure(data=[go.Pie(labels=meses, values=cantidad_ventas, title='Cantidades Pedidas por Mes de los Productos')])

# Muestra el gráfico
fig.show()

VENTAS TOTALES POR MES

In [198]:
import plotly.graph_objects as go
fig = go.Figure()

# Agrego al objeto figura un grafico de barra conn los valores extraidos previamente
fig.add_trace(go.Bar(
        x= meses,
        y= totales_ventas,
        yaxis="y",
        name="Monto",
        marker = dict(color= totales_ventas, colorscale = 'RdYlGn',colorbar = dict(title = "Total de ventas")),
        # marker = dict(color= 'Light Blue') ,

        text = totales_ventas,
        textposition = 'auto'
))


# Configuro titulos y ejes del grafico
fig.update_layout(
    title="Monto vendido por mes",
    xaxis=dict(title="Meses"),
    yaxis=dict(title="Monto", side="left")
)

# Muestro el grafico
fig.show()

Mes destacado a la venta

In [199]:
totales_ventas = pd.Series(totales_ventas) # conversion de lista a serie de PD

posicion_maximo = totales_ventas.idxmax()  # Posición en la lista
mes_mayor_venta = meses[posicion_maximo]     # Obtiene el Nombre del mes segun la posicion del valor de maximo
mayor_venta = totales_ventas[posicion_maximo]
mes_mayor_venta, mayor_venta

('Diciembre', 4619297)

Variacion de las ventas en los diferentes Meses

In [200]:
variacion_entre_meses = []

variacion_entre_meses = totales_ventas.diff()   # Calcula la diferencia con el mes anterior
variacion_entre_meses [0]= 0   # como la referencia inical es enero se asigna un valor 0

variacion_entre_meses

0           0.0
1      388146.0
2      605582.0
3      586996.0
4     -251475.0
5     -566291.0
6       68606.0
7     -402487.0
8     -145596.0
9     1638068.0
10    -537975.0
11    1420388.0
dtype: float64

In [201]:
fig = go.Figure()

# Agrego al objeto figura un grafico de barra conn los valores extraidos previamente
fig.add_trace(go.Bar(
        x= meses,
        y=  variacion_entre_meses,
        yaxis="y",
        
        marker = dict(color= totales_ventas, colorscale = 'RdYlGn'),
        # marker = dict(color= 'Light Blue') ,

        text = totales_ventas,
        textposition = 'auto'
))

fig.add_trace(go.Scatter(
        x = meses,
        y= variacion_entre_meses,
        
        #suaviza la linea roja
        line = dict(color= 'Red',shape='spline',smoothing=1.3,width=3),
        yaxis = 'y2'
        )
)

# Configuro titulos y ejes del grafico
fig.update_layout(
    
    xaxis=dict(title="Meses"),
    yaxis=dict(title="Unidades vendidas", side="left")
)

fig.update_layout(
    title = 'Diferencias de ventas tomando como referencia el mes anterior',
    xaxis = dict(title='Meses'),
    yaxis = dict(title='Diferencia Monto', side='left'),
    # esto es el titulo a la derecha
    yaxis2 = dict(title='', overlaying = 'y', side='right'),

    # el piso de 0 esta bien ubicado
    yaxis_range = [min(variacion_entre_meses)*1.1, variacion_entre_meses.max()*1.1],
    yaxis2_range = [min(variacion_entre_meses)*1.1, variacion_entre_meses.max()*1.1]
)
# Muestro el grafico
fig.show()

Calcular la diferencia de venta en relacion al promedio anual

In [202]:
diferencia_venta_promedio = []


promedio_anual_venta = int(sum(totales_ventas)/12)   # promedio anual de venta en 12 meses

promedio_anual_venta_lista = [promedio_anual_venta] * 12

for valor_mes in totales_ventas:

    diferencia_venta_promedio.append(int( valor_mes - promedio_anual_venta))   # diferencia mes a mes entre lo ventidi y el promedio mensual

diferencia_venta_promedio, promedio_anual_venta


([-1059001,
  -670855,
  -65273,
  521723,
  270248,
  -296043,
  -227437,
  -629924,
  -775520,
  862548,
  324573,
  1744961],
 2874336)

In [203]:
fig = go.Figure()

# Grafico de barras
fig.add_trace(go.Bar(
        x= meses,
        y=  diferencia_venta_promedio,
        
        marker = dict(color= totales_ventas, colorscale = 'RdYlGn'),
        text = diferencia_venta_promedio,
        textposition = 'auto'
))

# Linea 
fig.add_trace(go.Scatter(
        x = meses,
        y = diferencia_venta_promedio,
        
        #suaviza la linea roja
        line = dict(color= 'Red',shape='spline',smoothing=1.3,width=3),
        
))


# Configuro titulos y ejes del grafico

fig.update_layout(
    title = 'Diferencias de ventas entre meses con respecto al promedio anual',
    xaxis = dict(title='Meses'),
    yaxis = dict(title='Diferencia', side='left'),
    # esto es el titulo a la derecha
    yaxis2 = dict( overlaying = 'y'),
    
    # rango del eje y con un 10% mas del min y max valor
    yaxis_range = [min(diferencia_venta_promedio)*1.1, max(diferencia_venta_promedio) * 1.1],
    yaxis2_range = [min(diferencia_venta_promedio)*1.1, max(diferencia_venta_promedio) * 1.1],
    
)
# Muestro el grafico
fig.show()

2. Optimización de la publicidad y patrón de ventas por hora:
● ¿Cuál es el momento ideal del día para presentar la publicidad y aumentar la
probabilidad de compra?
● ¿Cómo cambian los patrones de ventas por hora a lo largo del año?
● ¿Hay modificaciones en los patrones de ventas durante las horas de mayor actividad
en los distintos meses?


In [204]:
# funcion para separa fecha y hora de los pedidos para poder analizarlos
def separ_datos(datos):
    datos['Fecha de Pedido']=pd.to_datetime(dataframe_mes['Fecha de Pedido']) #transformo datos de fecha en un formato legible por pandas
    datos['Hora']=datos['Fecha de Pedido'].dt.hour
    datos['Dia']=datos['Fecha de Pedido'].dt.day
    return datos


In [205]:
Horas = list(range(24)) # defino rango horario

fig = go.Figure()

for mes in meses:
  dataframe_mes = dataframes[mes]

  separ_datos(dataframe_mes) #llamo a la funcion para separa dia y hora
  grupos=dataframe_mes.groupby('Hora')['Cantidad Pedida'].sum() #agrupo por hora y sumo la cantidad pedida por hora

  hora_max_venta = grupos.idxmax()  # Obtiene la hora de mayor venta en el mes actual

# grafico, por cada iteracion del ciclo for (cada mes), me genera una linea en mi grafico identificada con distintos colores por cada mes
  fig.add_trace(go.Scatter(
    x = Horas,
    y = grupos,
    mode = 'lines',
    name = mes,
    line = dict(shape='spline', smoothing=1.3, width=2),
    text=f'Hora máxima de venta: {hora_max_venta}',  # Agrega el texto con la hora máxima de venta en el mes actual
    ))

fig.update_layout(
    title= "Unidades vendidas",
    xaxis=dict(title="Hora", tickvals= Horas),
    yaxis=dict(title="unidades vendidas", side="left"))
#imprime un solo grafico con cada uno de los meses del ciclo for anterior
fig.show()


D:\Temp\ipykernel_1168\1408740760.py:3: UserWarning:

Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.



D:\Temp\ipykernel_1168\1408740760.py:3: UserWarning:

Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.

D:\Temp\ipykernel_1168\1408740760.py:3: UserWarning:

Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.

D:\Temp\ipykernel_1168\1408740760.py:3: UserWarning:

Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.

D:\Temp\ipykernel_1168\1408740760.py:3: UserWarning:

Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.

D:\Temp\ipykernel_1168\1408740760.py:3: UserWarning:

Could not infer format, so each element wi

3. Distribución de ventas por ubicación:
● ¿En qué ciudades se han registrado las mayores ventas?
● ¿Cómo se comparan las ventas en diferentes estados o regiones?
● ¿Existe variación en las ventas por estado a lo largo de los meses?

CIUDADES CON MAYOR  VENTA

In [206]:
## se procede a desglosar la columna direccion de Envio

meses = ['Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio', 'Julio', 'Agosto', 'Septiembre', 'Octubre', 'Noviembre', 'Diciembre']

df_ventastotal_desglosado={}
for mes in meses:

  dataframe_mes = dataframes[mes]
  direccion_df = dataframe_mes['Dirección de Envio'].str.split(', ',expand=True) # se separa los datos de la columna
  direccion_df.columns = ['Calle','Ciudad','PO BOX']# se crea las columnas 
  direccion_df[['Estado', 'Cod Postal']] = direccion_df['PO BOX'].str.split(' ',expand=True)# 
  df_ventas_desglosado = pd.concat([dataframe_mes,direccion_df],axis=True)# se concatenan los dataframes
  df_ventastotal_desglosado[mes]=df_ventas_desglosado # se agrega al dataframe total deglosado
  
 

df_ventastotal_desglosado["Febrero"]


ID de Pedido                  Producto  Cantidad Pedida  \
0           150502                    iPhone                1   
1           150503     AA Batteries (4-pack)                1   
2           150504    27in 4K Gaming Monitor                1   
3           150505  Lightning Charging Cable                1   
4           150506     AA Batteries (4-pack)                2   
...            ...                       ...              ...   
12031       162004  Apple Airpods Headphones                1   
12032       162005    AAA Batteries (4-pack)                2   
12033       162006      USB-C Charging Cable                1   
12034       162007      USB-C Charging Cable                1   
12035       162008          27in FHD Monitor                1   

       Precio Unitario     Fecha de Pedido  \
0               700.00 2019-02-18 01:35:00   
1                 3.84 2019-02-13 07:24:00   
2               389.99 2019-02-18 09:46:00   
3                14.95 2019-02-02 16:47:00   
4                 3.84 2019-02-28 20:32:00   
...                ...                 ...   
12031           150.00 2019-02-12 22:02:00   
12032             2.99 2019-02-04 20:44:00   
12033            11.95 2019-02-24 06:31:00   
12034            11.95 2019-02-24 19:09:00   
12035           149.99 2019-02-26 17:15:00   

                            Dirección de Envio  Total venta  Hora  Dia  \
0            866 Spruce St, Portland, ME 04101       700.00     1   18   
1          18 13th St, San Francisco, CA 94016         3.84     7   13   
2           52 6th St, New York City, NY 10001       389.99     9   18   
3             129 Cherry St, Atlanta, GA 30301        14.95    16    2   
4            548 Lincoln St, Seattle, WA 98101         7.68    20   28   
...                                        ...          ...   ...  ...   
12031   227 Church St, San Francisco, CA 94016       150.00    22   12   
12032  417 Jefferson St, Los Angeles, CA 90001         5.98    20    4   
12033            498 8th St, Atlanta, GA 30301        11.95     6   24   
12034             715 7th St, Dallas, TX 75001        11.95    19   24   
12035       677 West St, Los Angeles, CA 90001       149.99    17   26   

                  Calle         Ciudad    PO BOX Estado Cod Postal  
0         866 Spruce St       Portland  ME 04101     ME      04101  
1            18 13th St  San Francisco  CA 94016     CA      94016  
2             52 6th St  New York City  NY 10001     NY      10001  
3         129 Cherry St        Atlanta  GA 30301     GA      30301  
4        548 Lincoln St        Seattle  WA 98101     WA      98101  
...                 ...            ...       ...    ...        ...  
12031     227 Church St  San Francisco  CA 94016     CA      94016  
12032  417 Jefferson St    Los Angeles  CA 90001     CA      90001  
12033        498 8th St        Atlanta  GA 30301     GA      30301  
12034        715 7th St         Dallas  TX 75001     TX      75001  
12035       677 West St    Los Angeles  CA 90001     CA      90001  

[11986 rows x 14 columns]

VENTAS POR ESTADOS

In [207]:
dataframe_total = pd.DataFrame()
cont=0
for mes in meses:
  df_mes=df_ventastotal_desglosado[mes] # se optiene el datafram de cada mes
  cont=cont+1
  df_mes["Mes"]=mes #agrego una nueva columna con el mes correspondiente
  df_mes["N-mes"]=cont #agrego otra columna con el numero de mes
  dataframe_total = pd.concat([dataframe_total, df_mes], ignore_index=True)#  se concatenan los dataframe
  
dataframe_total

ID de Pedido                    Producto  Cantidad Pedida  \
0            141234                      iPhone                1   
1            141235    Lightning Charging Cable                1   
2            141236            Wired Headphones                2   
3            141237            27in FHD Monitor                1   
4            141238            Wired Headphones                1   
...             ...                         ...              ...   
185945       319666    Lightning Charging Cable                1   
185946       319667       AA Batteries (4-pack)                2   
185947       319668             Vareebadd Phone                1   
185948       319669            Wired Headphones                1   
185949       319670  Bose SoundSport Headphones                1   

        Precio Unitario     Fecha de Pedido  \
0                700.00 2019-01-22 21:25:00   
1                 14.95 2019-01-28 14:15:00   
2                 11.99 2019-01-17 13:33:00   
3                149.99 2019-01-05 20:33:00   
4                 11.99 2019-01-25 11:59:00   
...                 ...                 ...   
185945            14.95 2019-12-11 20:58:00   
185946             3.84 2019-12-01 12:01:00   
185947           400.00 2019-12-09 06:43:00   
185948            11.99 2019-12-03 10:39:00   
185949            99.99 2019-12-21 21:45:00   

                            Dirección de Envio  Total venta  Hora  Dia  \
0              944 Walnut St, Boston, MA 02215       700.00    21   22   
1             185 Maple St, Portland, OR 97035        14.95    14   28   
2        538 Adams St, San Francisco, CA 94016        23.98    13   17   
3           738 10th St, Los Angeles, CA 90001       149.99    20    5   
4                387 10th St, Austin, TX 73301        11.99    11   25   
...                                        ...          ...   ...  ...   
185945  14 Madison St, San Francisco, CA 94016        14.95    20   11   
185946    549 Willow St, Los Angeles, CA 90001         7.68    12    1   
185947        273 Wilson St, Seattle, WA 98101       400.00     6    9   
185948          778 River St, Dallas, TX 75001        11.99    10    3   
185949  747 Chestnut St, Los Angeles, CA 90001        99.99    21   21   

                  Calle         Ciudad    PO BOX Estado Cod Postal        Mes  \
0         944 Walnut St         Boston  MA 02215     MA      02215      Enero   
1          185 Maple St       Portland  OR 97035     OR      97035      Enero   
2          538 Adams St  San Francisco  CA 94016     CA      94016      Enero   
3           738 10th St    Los Angeles  CA 90001     CA      90001      Enero   
4           387 10th St         Austin  TX 73301     TX      73301      Enero   
...                 ...            ...       ...    ...        ...        ...   
185945    14 Madison St  San Francisco  CA 94016     CA      94016  Diciembre   
185946    549 Willow St    Los Angeles  CA 90001     CA      90001  Diciembre   
185947    273 Wilson St        Seattle  WA 98101     WA      98101  Diciembre   
185948     778 River St         Dallas  TX 75001     TX      75001  Diciembre   
185949  747 Chestnut St    Los Angeles  CA 90001     CA      90001  Diciembre   

        N-mes  
0           1  
1           1  
2           1  
3           1  
4           1  
...       ...  
185945     12  
185946     12  
185947     12  
185948     12  
185949     12  

[185950 rows x 16 columns]

In [208]:
col_filtradas_ciudad = ['Ciudad','Cantidad Pedida','N-mes','Mes']
df_total_reducido_ciudad = dataframe_total[col_filtradas_ciudad] #filtrado de del dataframe con las columnas necesarias

df_suma_cantidad_anual = df_total_reducido_ciudad.groupby(['Mes',"N-mes","Ciudad"])['Cantidad Pedida'].sum().reset_index()
dfordenado_c=df_suma_cantidad_anual.sort_values("N-mes")#ordeno por mes


dfordenado_c

Mes  N-mes         Ciudad  Cantidad Pedida
32      Enero      1  New York City             1497
28      Enero      1         Austin              580
29      Enero      1         Boston             1171
30      Enero      1         Dallas              879
31      Enero      1    Los Angeles             1716
..        ...    ...            ...              ...
22  Diciembre     12    Los Angeles             4373
21  Diciembre     12         Dallas             2300
18  Diciembre     12        Atlanta             2224
23  Diciembre     12  New York City             3828
19  Diciembre     12         Austin             1520

[108 rows x 4 columns]

In [209]:
df_total_ciudad = dfordenado_c.groupby("Ciudad")['Cantidad Pedida'].sum().reset_index()

df_total_ciudad

Ciudad  Cantidad Pedida
0        Atlanta            16602
1         Austin            11153
2         Boston            22528
3         Dallas            16730
4    Los Angeles            33289
5  New York City            27932
6       Portland            14053
7  San Francisco            50239
8        Seattle            16553

In [210]:
import pandas as pd
import plotly.express as px
fig = px.bar(df_total_ciudad, x='Ciudad', y='Cantidad Pedida', 
             color="Ciudad",barmode="stack",
             title="VENTAS total POR Ciudades",text_auto="Cantidad Pedida")

fig.show()

In [211]:
ciudad=list(df_total_ciudad["Ciudad"])
ventas_total_ciudad=list(df_total_ciudad["Cantidad Pedida"])
ciudad,ventas_total_ciudad

(['Atlanta',
  'Austin',
  'Boston',
  'Dallas',
  'Los Angeles',
  'New York City',
  'Portland',
  'San Francisco',
  'Seattle'],
 [16602, 11153, 22528, 16730, 33289, 27932, 14053, 50239, 16553])

In [212]:
import plotly.graph_objects as go
fig = go.Figure(data=[go.Pie(labels=ciudad, values=ventas_total_ciudad, title="UNIDADES VENDIDAS POR CIUDADES ",textinfo='label+percent',insidetextorientation='radial')])

# Muestra el gráfico
fig.show()

In [213]:
# Filtrado de Estados
col_filtradas_estado = ['Estado','Cantidad Pedida','N-mes','Mes']
df_total_reducido_estado = dataframe_total[col_filtradas_estado] #filtrado de del dataframe con las columnas necesarias
df_total_reducido_estado

Estado  Cantidad Pedida  N-mes        Mes
0          MA                1      1      Enero
1          OR                1      1      Enero
2          CA                2      1      Enero
3          CA                1      1      Enero
4          TX                1      1      Enero
...       ...              ...    ...        ...
185945     CA                1     12  Diciembre
185946     CA                2     12  Diciembre
185947     WA                1     12  Diciembre
185948     TX                1     12  Diciembre
185949     CA                1     12  Diciembre

[185950 rows x 4 columns]

In [214]:
#Suma de las cantidades pedidas por Estados
df_suma_cantidad_anual = df_total_reducido_estado.groupby(['Mes',"N-mes","Estado"])['Cantidad Pedida'].sum().reset_index()
dfordenado=df_suma_cantidad_anual.sort_values("N-mes")#ordeno por mes
dfordenado


Mes  N-mes Estado  Cantidad Pedida
29      Enero      1     OR              553
25      Enero      1     GA              877
26      Enero      1     MA             1171
27      Enero      1     ME              138
28      Enero      1     NY             1497
..        ...    ...    ...              ...
20  Diciembre     12     NY             3828
21  Diciembre     12     OR             1507
22  Diciembre     12     TX             3820
23  Diciembre     12     WA             2202
17  Diciembre     12     GA             2224

[96 rows x 4 columns]

In [215]:
df_total_estados = df_total_reducido_estado.groupby("Estado")['Cantidad Pedida'].sum().reset_index()

df_total_estados

Estado  Cantidad Pedida
0     CA            83528
1     GA            16602
2     MA            22528
3     ME             2750
4     NY            27932
5     OR            11303
6     TX            27883
7     WA            16553

In [216]:
import pandas as pd
import plotly.express as px
fig = px.bar(df_total_estados, x='Estado', y='Cantidad Pedida', 
             color="Estado",barmode="stack",
             title="VENTAS total POR ESTADOS",text_auto="Cantidad Pedida")

fig.show()


GRAFICOS DE BARRAS APILADAS

In [217]:
import pandas as pd
import plotly.express as px
fig = px.bar(dfordenado, x='Estado', y='Cantidad Pedida', 
             color="Mes",barmode="stack",
             title="VENTAS POR ESTADOS",text_auto="Cantidad Pedida")

fig.show()

CANTIDAD VENDIDAS POR ESTADOS A LO LARGO DE LOS MESES

In [218]:
fig = go.Figure()


for mes in meses:
  dataframe_mes = df_ventastotal_desglosado[mes]
  

  Ventas_por_mes= dataframe_mes.groupby('Estado')['Cantidad Pedida'].sum()
  Ventas_por_mes1=pd.DataFrame(Ventas_por_mes)
  Ventas_por_mes1.reset_index(inplace=True)
  Estado=Ventas_por_mes1["Estado"]
  


  # Agrego al objeto figura un grafico de de linea con las horas y las ventas por hora
  fig.add_trace(go.Scatter(

      x=Estado,
      y=Ventas_por_mes,
      mode='lines',
      name=mes,

      line=dict(shape='spline', smoothing=1.3, width=2)))


  # Configuro los ejes y los títulos
fig.update_layout(

    title="Unidades vendidas por Estado",
    xaxis=dict(title="Estado", tickvals=Estado),  # Para que en el eje x se muestren todos los valores uso tickvals=Horas
    yaxis=dict(title="Unidades vendidas", side="left"))


fig.show()

GRAFICO DE CANTIDAD VENDIDA POR ESTADOS 

In [219]:
import plotly.express as px

fig = px.choropleth(df_total_estados, locations='Estado', locationmode='USA-states',color='Cantidad Pedida',
                           color_continuous_scale="RdYlGn",
                           scope="usa",
                          title= 'Cantidad vendidad por Estados'
                          )

fig.show()

4. Análisis del producto más vendido:
● ¿Cuál es el producto más vendido en general y en cada mes?
● ¿Qué factores crees que han contribuido al éxito de ese producto en particular?


In [220]:
#  Generar un dataframe general con todo el año
dataframe_total = pd.DataFrame()
for mes in meses:
  dataframe_total = pd.concat([dataframe_total, dataframes[mes]], ignore_index=True)

dataframe_total


ID de Pedido                    Producto  Cantidad Pedida  \
0            141234                      iPhone                1   
1            141235    Lightning Charging Cable                1   
2            141236            Wired Headphones                2   
3            141237            27in FHD Monitor                1   
4            141238            Wired Headphones                1   
...             ...                         ...              ...   
185945       319666    Lightning Charging Cable                1   
185946       319667       AA Batteries (4-pack)                2   
185947       319668             Vareebadd Phone                1   
185948       319669            Wired Headphones                1   
185949       319670  Bose SoundSport Headphones                1   

        Precio Unitario     Fecha de Pedido  \
0                700.00 2019-01-22 21:25:00   
1                 14.95 2019-01-28 14:15:00   
2                 11.99 2019-01-17 13:33:00   
3                149.99 2019-01-05 20:33:00   
4                 11.99 2019-01-25 11:59:00   
...                 ...                 ...   
185945            14.95 2019-12-11 20:58:00   
185946             3.84 2019-12-01 12:01:00   
185947           400.00 2019-12-09 06:43:00   
185948            11.99 2019-12-03 10:39:00   
185949            99.99 2019-12-21 21:45:00   

                            Dirección de Envio  Total venta  Hora  Dia  
0              944 Walnut St, Boston, MA 02215       700.00    21   22  
1             185 Maple St, Portland, OR 97035        14.95    14   28  
2        538 Adams St, San Francisco, CA 94016        23.98    13   17  
3           738 10th St, Los Angeles, CA 90001       149.99    20    5  
4                387 10th St, Austin, TX 73301        11.99    11   25  
...                                        ...          ...   ...  ...  
185945  14 Madison St, San Francisco, CA 94016        14.95    20   11  
185946    549 Willow St, Los Angeles, CA 90001         7.68    12    1  
185947        273 Wilson St, Seattle, WA 98101       400.00     6    9  
185948          778 River St, Dallas, TX 75001        11.99    10    3  
185949  747 Chestnut St, Los Angeles, CA 90001        99.99    21   21  

[185950 rows x 9 columns]

In [221]:
# Eliminar campos no deseados
columnas_filtradas = ['Producto','Cantidad Pedida','Precio Unitario']
df_total_reducido = dataframe_total[columnas_filtradas]
df_total_reducido

Producto  Cantidad Pedida  Precio Unitario
0                           iPhone                1           700.00
1         Lightning Charging Cable                1            14.95
2                 Wired Headphones                2            11.99
3                 27in FHD Monitor                1           149.99
4                 Wired Headphones                1            11.99
...                            ...              ...              ...
185945    Lightning Charging Cable                1            14.95
185946       AA Batteries (4-pack)                2             3.84
185947             Vareebadd Phone                1           400.00
185948            Wired Headphones                1            11.99
185949  Bose SoundSport Headphones                1            99.99

[185950 rows x 3 columns]

In [222]:
 # genera un df agrupado por el nomnbbre del producto y suma todos los pedidos en el campo Cantidad Pedida
df_suma_cantidad_anual = df_total_reducido.groupby('Producto')['Cantidad Pedida'].sum().reset_index()
df_suma_cantidad_anual

Producto  Cantidad Pedida
0                 20in Monitor             4129
1       27in 4K Gaming Monitor             6244
2             27in FHD Monitor             7550
3       34in Ultrawide Monitor             6199
4        AA Batteries (4-pack)            27635
5       AAA Batteries (4-pack)            31017
6     Apple Airpods Headphones            15661
7   Bose SoundSport Headphones            13457
8                Flatscreen TV             4819
9                 Google Phone             5532
10                    LG Dryer              646
11          LG Washing Machine              666
12    Lightning Charging Cable            23217
13          Macbook Pro Laptop             4728
14             ThinkPad Laptop             4130
15        USB-C Charging Cable            23975
16             Vareebadd Phone             2068
17            Wired Headphones            20557
18                      iPhone             6849

In [223]:
import plotly.express as px

In [224]:
# grafico de torta

fig = px.pie(df_ventas, values='Cantidad Pedida', names='Producto', title='Cantidades Pedidas por Producto en el año')

# Muestra el gráfico
fig.show()

Producto mas Vendido en todo el año

In [225]:
index_producto_mas_pedido = df_suma_cantidad_anual['Cantidad Pedida'].idxmax()
producto_mas_pedido_anual = df_suma_cantidad_anual.at[index_producto_mas_pedido, 'Producto']
cantidad_mas_pedida_anual = df_suma_cantidad_anual.at[index_producto_mas_pedido, 'Cantidad Pedida']
producto_mas_pedido_anual , cantidad_mas_pedida_anual

('AAA Batteries (4-pack)', 31017)

Productos mas Pedidos cada mes

In [226]:
# Comprobacion de cantidad de registros en cada mes, la sumatoria de todos y el total acumulado
# De esta manera se determina que estan bien organizados
producto_mas_pedido_mes =[]
cantidad_mas_pedida_mes = []
for mes in meses:
  dataframe_mes = dataframes[mes].groupby('Producto')['Cantidad Pedida'].sum().reset_index()  # genera un df agrupado por el nomnbbre del producto y suma todos los pedidos en el campo Cantidad Pedida
  index_producto_mas_pedido = dataframe_mes['Cantidad Pedida'].idxmax() # busca la ubicacion del mayor valor

  # dada la referencia del maximo valor agrega a la lista correspondiente al mes el nombre y cantidad del producto mas vendido
  producto_mas_pedido_mes.append(dataframe_mes.at[index_producto_mas_pedido, 'Producto'])
  cantidad_mas_pedida_mes.append(dataframe_mes.at[index_producto_mas_pedido, 'Cantidad Pedida'])

producto_mas_pedido_mes,cantidad_mas_pedida_mes



(['AAA Batteries (4-pack)',
  'AAA Batteries (4-pack)',
  'AAA Batteries (4-pack)',
  'AAA Batteries (4-pack)',
  'AAA Batteries (4-pack)',
  'AAA Batteries (4-pack)',
  'AAA Batteries (4-pack)',
  'AAA Batteries (4-pack)',
  'AAA Batteries (4-pack)',
  'AAA Batteries (4-pack)',
  'AAA Batteries (4-pack)',
  'AAA Batteries (4-pack)'],
 [1596, 1975, 2478, 2936, 2931, 2160, 2366, 2019, 1927, 3362, 3027, 4240])

Grafico de productos vendidos y cantidades en los meses

In [227]:
fig = go.Figure()

# Grafico de barras
fig.add_trace(go.Bar(
        x= meses,
        y= cantidad_mas_pedida_mes,
        yaxis="y",

        marker = dict(color= cantidad_mas_pedida_mes, colorscale = 'RdYlGn',colorbar = dict(title = "Cambios")),
        text = producto_mas_pedido_mes,
        textposition = 'auto'
))


# Configuro titulos y ejes del grafico

fig.update_layout(
    title = 'Productos mas vendidos en cada mes',
    xaxis = dict(title='Meses'),
    yaxis = dict(title='Cantidad', side='left'),

    # rango del eje y con un 10% mas del min y max valor
    yaxis_range = [0, max(cantidad_mas_pedida_mes) * 1.1],


)
# Muestro el grafico
fig.show()

In [228]:
# Cantidad de Productos mas vendidos
ubicacion_del_producto = df_ventas['Cantidad Pedida'].idxmax()
cantidad_mas_vendido = df_ventas['Cantidad Pedida'][ubicacion_del_producto]

cantidad_mas_vendido

6

Factores  que han contribuido al éxito de ese producto en particular

In [229]:
index_menor_precio = dataframe_total ['Precio Unitario'].idxmin()
menor_precio = dataframe_total['Precio Unitario'][index_menor_precio]
producto_menor_precio = dataframe_total['Producto'][index_menor_precio]
menor_precio,producto_menor_precio

(2.99, 'AAA Batteries (4-pack)')

CONCLUSIONES DEL PUNTO 4

A - Producto mas vendido en el año es  AAA Batteries (4-pack) con 31017 ventas- (Grafico torta)

B - Producto mas vendido en cada mes es  AAA Batteries (4-pack) (Grafico barras)


C - No hay nada representativo en el porque de la mayor venta anual y en todos los meses del mismo producto, pero se pueden tomar las siguientes opiniones:

-- Por ser un producto economico, es el mas economico entre los demas.
-- Un producto de desgaste, cuando se agota su carga deben ser desechadas y volver a adquirir, o sea el tiempo de vida es menor al de un monitor
-- Un producto de uso masivo, es normal tener varios controles remotos o pequeños artefactos en el hogar que usen pares de pilas, por ende la demanda es mayor

5. Tendencia ventas:
● ¿Existe alguna tendencia o patrón en las ventas que se repita a lo largo de los días
en los diferentes meses?
● ¿Cómo varían las ventas a lo largo de los diferentes días de la semana?
● ¿Se observa alguna diferencia significativa en las ventas entre los días laborables y
los fines de semana?

In [230]:

# Pasar los dataframe a dias


df_por_dias = dataframe_total # realizo una copia el df con toda la informacion del año para no modicfiocar el original

# df_por_dias['Fecha de Pedido'] = pd.to_datetime(df_por_dias['Fecha de Pedido'])

df_por_dias['solo_dia']=df_por_dias['Fecha de Pedido'].dt.date

df_por_dias

# Agrupo por dias

df_suma_diaria = df_por_dias.groupby('solo_dia')['Cantidad Pedida'].sum().reset_index() 

df_suma_diaria

df_suma_diaria['Cantidad Pedida'].max()


# Separar los dias de los meses


meses = ['Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio', 'Julio', 'Agosto', 'Septiembre', 'Octubre', 'Noviembre', 'Diciembre']

# Separar el número de día y el mes
df_dia_mes_valor=df_suma_diaria

df_dia_mes_valor['Dia'] = df_dia_mes_valor['solo_dia'].apply(lambda x: x.day)
df_dia_mes_valor['Mes'] = df_dia_mes_valor['solo_dia'].apply(lambda x: x.month)
df_dia_mes_valor['Anio'] = df_dia_mes_valor['solo_dia'].apply(lambda x: x.year)

colores=['Yellow','Gold','Orange','Red','Violet','Purple','Blue','Teal','Lime','Green','Brown','Grey']

df_dia_mes_valor.drop(df_dia_mes_valor[(df_dia_mes_valor['Anio']==2020)].index,inplace=True) # se borra el primer dia del año 2020
# Iterar por cada mes único
meses_unicos = df_dia_mes_valor['Mes'].unique()  # genere un array con los numeros de los meses
fig = go.Figure()

for mes,color in zip(meses_unicos,colores):
    # Filtrar el DataFrame por mes
    
    df_mes_2019 = df_dia_mes_valor[(df_dia_mes_valor['Anio']== 2019) & (df_dia_mes_valor['Mes'] == mes)]
    # Obtener los valores de niveles para cada día en una lista
    niveles_por_dia = df_mes_2019['Cantidad Pedida'].tolist()
    indice_picos=np.argmax(niveles_por_dia)


    fig.add_trace(go.Scatter(
        x=df_mes_2019['Dia'],
        y=df_mes_2019['Cantidad Pedida'], # cantidad vendidas por mes
        mode='lines',
        name=f' {meses[mes-1]}',
        marker_color=color

        )
    )
    fig.add_trace(go.Scatter(
        x=[df_mes_2019['Dia'].iloc[indice_picos]],
        y=[niveles_por_dia[indice_picos]],
        mode='markers',
        marker=dict(size=8,color='red'),
        showlegend=False
    ))


fig.update_layout(
    title='Ventas Diarias por Mes',
    xaxis=dict(title='Días',tickvals= df_mes_2019['Dia']),
    yaxis=dict(title='Ventas', side='left'),
    showlegend=True,
)
fig.show()

# los patrones que encuentro en el grafico es que en enero hay pocas ventas (supongo que por el gasto que hubo en el mes anterior en diciembre)
# $ los patrones que se encuentran es en los picos son siempre despues de la mitaad del mes

In [231]:
# mostrar por cada 7 dias del mes las ventas de la semana

from calendar import month_name  # Importa el nombre del mes


df_semana=df_por_dias
df_semana['Anio'] = df_semana['solo_dia'].apply(lambda x: x.year)
df_semana['Mes'] = df_semana['solo_dia'].apply(lambda x: x.month)
df_semana['Dia'] = df_semana['solo_dia'].apply(lambda x: x.day)

df_semana.drop(df_semana[(df_semana['Anio']==2020)].index,inplace=True) # se borra el primer dia del año 2020

df_semana= df_semana.groupby(['Dia','Mes'])['Cantidad Pedida'].sum().reset_index() #agrupar por año,mes y dia de semana la suma total de las ventas
df_semana_filtrada=df_semana


# Función para clasificar los días en semanas (primera, segunda, tercera, última)
def clasificar_semana(dia):
    if dia <= 7:
        return 'Primera'
    elif dia <= 14:
        return 'Segunda'
    elif dia <= 21:
        return 'Tercera'
    else:
        return 'Última'

# Aplicar la función para crear una nueva columna 'Semana'
df_semana_filtrada['Semana'] = df_semana['Dia'].apply(clasificar_semana)

# Calcular la suma total de las cosas vendidas en cada semana
resumen_semanal = df_semana_filtrada.groupby(['Mes', 'Semana'])['Cantidad Pedida'].sum().reset_index()

# Agregar una columna con nombres de mes
resumen_semanal['Meses'] = resumen_semanal['Mes'].map(lambda x: month_name[x])
mesIngles=['January','February','March','April','May','June','July','August','September','October','November','December']
resumen_semanal['Meses'] = resumen_semanal['Meses'].replace(mesIngles, meses) # se reemplazan las palabras en ingles de los dias de la semana a español


# Crear el gráfico de barras con Plotly
fig = px.bar(resumen_semanal, x='Meses', y='Cantidad Pedida', color='Semana',
             title='Suma Total de Cantidad Pedida por Semana en Cada Mes',
             labels={'Cantidad Pedida': 'Cantidad Pedida Total'})

# Añade etiquetas a las barras
fig.update_traces(texttemplate='%{y}')
# Mostrar el gráfico me
fig.show()

In [232]:
# se realiza un grafico con las 4 smanas del mes

fig = px.bar(resumen_semanal, x='Semana', y='Cantidad Pedida', color='Cantidad Pedida', facet_col='Meses',
             category_orders={'Semana': ['Primera', 'Segunda', 'Tercera', 'Última']},
             labels={'Cantidad Pedida': 'Cantidad Vendida'},
             height=600, width=1400)  # Ajusta el alto y el ancho según tus preferencias

fig.update_layout(title='Ventas por Semana por Mes',
                  xaxis_title='Día de la Semana',
                  yaxis_title='Cantidad Vendida')

# Se personalizan las etiquetas de las facetas para eliminar "Mes="
fig.for_each_annotation(lambda a: a.update(text=a.text.split("=")[1]))
fig.update_xaxes(title_text='')

fig.show()

In [233]:
# grafico de torta sobre cada semana a lo largo del año

fig = px.pie(resumen_semanal, values='Cantidad Pedida', names='Semana', title='Cantidades Pedidas por Semana en el año')

# Muestra el gráfico
fig.show()

In [234]:
df_correccion=df_por_dias
df_correccion['Anio'] = df_correccion['solo_dia'].apply(lambda x: x.year)
df_correccion['Mes'] = df_correccion['solo_dia'].apply(lambda x: x.month)

mostrar = df_correccion[(df_correccion['Anio']== 2020)] # es para ver cuantos del 2020 hay



df_correccion['Dia_semana']=df_correccion['Fecha de Pedido'].dt.day_name() #pasa del datetime a los dias de la semana en ingles

df_correccion.drop(df_correccion[(df_correccion['Anio']==2020)].index,inplace=True) # se borra el primer dia del año 2020

df_correccion= df_correccion.groupby(['Anio','Mes','Dia_semana'])['Cantidad Pedida'].sum().reset_index() #agrupar por año,mes y dia de semana la suma total de las ventas

dias_semana=['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'] #Semana en ingles
nombres_dias_espanol = ['Lunes', 'Martes', 'Miércoles', 'Jueves', 'Viernes', 'Sábado', 'Domingo'] #Semana en español
meses_palabra=['Enero','Febrero','Marzo','Abril','Mayo','Junio','Julio','Agosto','Septiembre','Octubre','Noviembre','Diciembre'] # meses palabras

df_correccion['Dia_semana'] = df_correccion['Dia_semana'].replace(dias_semana, nombres_dias_espanol) # se reemplazan las palabras en ingles de los dias de la semana a español

colores=['yellow', 'orange', 'red', 'violet', 'blue', 'green', 'brown']



fig = go.Figure()

for dia,color in zip(nombres_dias_espanol,colores):
    df_dia=df_correccion[df_correccion['Dia_semana']==dia]
    fig.add_trace(go.Bar(
        x=df_dia['Mes'],
        y=df_dia['Cantidad Pedida'], # cantidad vendidas por mes
        name=dia,
        marker_color=color
        )
    )
    
fig.update_layout(
    barmode='stack',
    title='Cantidades pedidas por dia de la semana por mes durante todo el Año',
    xaxis=dict(title='Mes',tickvals= df_dia['Mes'], ticktext=meses_palabra,categoryorder='array', categoryarray=meses_palabra),  # Especificar el orden personalizado),
    yaxis=dict(title='Ventas', side='left'),
    legend=dict(title='Día de la Semana')
)

fig.update_traces(texttemplate='%{y}')

fig.show()




In [235]:
# gráfico de barras con una línea punteada en el medio que significa la media anual en base a la cantidad de pedidos mensuales.

df_diaSemana_mes=df_correccion

dataMes = {
    'Mes':['Enero','Febrero','Marzo','Abril','Mayo','Junio','Julio','Agosto','Septiembre','Octubre','Noviembre','Diciembre']
}

df_mes = pd.DataFrame(dataMes)

df_diaSemana_mes = df_diaSemana_mes.sort_values(by=['Mes'])

meses_nombre=['Enero','Febrero','Marzo','Abril','Mayo','Junio','Julio','Agosto','Septiembre','Octubre','Noviembre','Diciembre']

df_diaSemana_mes['Mes'] = df_diaSemana_mes['Mes'].map(lambda x: meses_nombre[x - 1])

fig = px.bar(df_diaSemana_mes, x='Dia_semana', y='Cantidad Pedida', color='Cantidad Pedida', facet_col='Mes',
             category_orders={'Dia_semana': ['Lunes', 'Martes', 'Miércoles', 'Jueves', 'Viernes', 'Sábado', 'Domingo']},
             labels={'Cantidad Pedida': 'Cantidad Pedida'},
             height=600, width=1800)
fig.update_layout(title='Ventas por Día de la Semana por Mes',
                  xaxis_title='Día de la Semana',
                  yaxis_title='Cantidad Vendida')

# Se personalizan las etiquetas de las facetas para eliminar "Mes="
fig.for_each_annotation(lambda a: a.update(text=a.text.split("=")[1]))
fig.update_xaxes(title_text='')

fig.show()


In [236]:
# mostrar en un grafico torta todos los dias de la semana cuántos pedidos se hicieron a lo largo del año

# Definir el orden de los días de la semana
orden_dias = ['Lunes', 'Martes', 'Miércoles', 'Jueves', 'Viernes', 'Sábado', 'Domingo']

# Ordenar el DataFrame en función del orden de los días de la semana
df_correccion['Dia_semana'] = pd.Categorical(df_correccion['Dia_semana'], categories=orden_dias, ordered=True)
df_correccion = df_correccion.sort_values(by='Dia_semana')



ventas_promedio_por_dia = df_correccion.groupby('Dia_semana')['Cantidad Pedida'].mean().reset_index() # ventas promedio por dia de semana anual
ventas_por_mes= df_correccion.groupby('Mes')['Cantidad Pedida'].sum().reset_index()
df_correccion
df_cantidadPedida_diaSemana= df_correccion.groupby('Dia_semana')['Cantidad Pedida'].sum().reset_index()
df_cantidadPedida_diaSemana

# grafico de torta

fig = px.pie(df_cantidadPedida_diaSemana, values='Cantidad Pedida', names='Dia_semana', title='Porcentajes de las Cantidades Pedidas por dia de la Semana en el año')

# Muestra el gráfico
fig.show()


D:\Temp\ipykernel_1168\3158288969.py:12: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

D:\Temp\ipykernel_1168\3158288969.py:15: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



In [237]:

totalesventas=df_cantidadPedida_diaSemana["Cantidad Pedida"]

variacion_entre_dias = []

variacion_entre_dias = totalesventas.diff()   # Calcula la diferencia con el mes anterior
variacion_entre_dias [0]= 0   # como la referencia inical es enero se asigna un valor 0

variacion_entre_dias


fig = go.Figure()

# Agrego al objeto figura un grafico de barra conn los valores extraidos previamente
fig.add_trace(go.Bar(
        x= orden_dias,
        y=  variacion_entre_dias,
        yaxis="y",
        name="",
        marker = dict(color= variacion_entre_dias, colorscale = 'RdYlGn',colorbar = dict(title = "Total de ventas")),
        # marker = dict(color= 'Light Blue') ,

        text = totalesventas,
        textposition = 'auto'
))

fig.add_trace(go.Scatter(
        x = orden_dias,
        y= variacion_entre_dias,
        name = 'Diferencia',
        #suaviza la linea roja
        line = dict(color= 'Red',shape='spline',smoothing=1.3,width=3),
        yaxis = 'y2'
        )
)

# Configuro titulos y ejes del grafico
fig.update_layout(
    title="Diferencia de Ventas entre dias ",
    xaxis=dict(title="Dias"),
    yaxis=dict(title="Unidades vendidas", side="left")
)

fig.update_layout(
    title = 'Diferencias de Cantidades pedidas tomando como referencia el dia Lunes',
    xaxis = dict(title='Dias de Semana'),
    yaxis = dict(title='Cantidad Pedida', side='left'),
    # esto es el titulo a la derecha
    yaxis2 = dict(title='Cantidad Pedida', overlaying = 'y', side='right'),

    # el piso de 0 esta bien ubicado
    yaxis_range = [min(variacion_entre_dias)*1.1, variacion_entre_dias.max()*1.1],
    yaxis2_range = [min(variacion_entre_dias)*1.1, variacion_entre_dias.max()*1.1]
)
# Muestro el grafico
fig.show()

In [238]:
# CALCULAR EN BASE A LA DIFERENCIA DELMINIMO Y EL MAXIMO

df_correccion
ventaDiaMax = df_correccion.groupby('Dia_semana')['Cantidad Pedida'].max().reset_index() # ventas promedio por dia de semana anual
ventaDiaMax["Cantidad Pedida"].max()

ventaDiaMax

ventaDiasSemana=df_correccion.groupby("Dia_semana")["Cantidad Pedida"].sum().reset_index()


ventaDiasSemanaMax= ventaDiasSemana["Cantidad Pedida"].max()
ventaDiasSemanaMin= ventaDiasSemana["Cantidad Pedida"].min()

ventaDiasSemanaMax
ventaDiasSemanaMin

# Calcula los porcentajes de ventas en relación con el mínimo y el máximo
ventaDiasSemana['Porcentaje Ventas'] = (ventaDiasSemana['Cantidad Pedida'] - ventaDiasSemanaMin) / (ventaDiasSemanaMax - ventaDiasSemanaMin) * 100

ventaDiasSemana

# Crea el gráfico de barras
fig = px.bar(ventaDiasSemana, x='Dia_semana', y='Porcentaje Ventas', text='Porcentaje Ventas',
             title='Porcentaje de Ventas por Día de la Semana en relacion con el mínimo y el máximo')

# Añade etiquetas a las barras
fig.update_traces(texttemplate='%{text:.2f}%')
fig.update_xaxes(title_text='Dias de la semana')

# Muestra el gráfico
fig.show()


D:\Temp\ipykernel_1168\2800875561.py:4: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

D:\Temp\ipykernel_1168\2800875561.py:9: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



In [239]:
# gráfico de barras con una línea punteada en el medio que significa la media anual en base a la cantidad de pedidos mensuales.


fig = go.Figure()

# Se observa alguna diferencia significativa en las ventas entre los días laborables y los fines de semana?
df_DiasLaborales_Finde=df_correccion

dias_laborales={
    'Lunes':'Laboral',
    'Martes':'Laboral',
    'Miércoles':'Laboral',
    'Jueves':'Laboral',
    'Viernes':'Laboral',
    'Sábado':'No Laboral',
    'Domingo':'No Laboral'
}

df_DiasLaborales_Finde['tipo dia']= df_DiasLaborales_Finde['Dia_semana'].map(dias_laborales)

# Filtrar los datos para obtener solo los días laborales.
dias_laborales_df = df_DiasLaborales_Finde[df_DiasLaborales_Finde['tipo dia'] == 'Laboral']

# Agrupar por mes y sumar la cantidad de cosas vendidas en días laborales.
dias_laborales_por_mes = dias_laborales_df.groupby('Mes')['Cantidad Pedida'].sum().reset_index()

# Filtrar los datos para obtener solo los días no laborales.
no_laborales_df = df_DiasLaborales_Finde[df_DiasLaborales_Finde['tipo dia'] == 'No Laboral']

# Agrupar por mes y sumar la cantidad de cosas vendidas en días no laborales.
no_laborales_por_mes = no_laborales_df.groupby('Mes')['Cantidad Pedida'].sum().reset_index()

# Barras de días laborales.
fig.add_trace(go.Bar(
    x=dias_laborales_por_mes['Mes'],
    y=dias_laborales_por_mes['Cantidad Pedida'],
    name='Laborales',
    text=dias_laborales_por_mes['Cantidad Pedida'],
    marker_color='blue'
))

# Barras de días no laborales.
fig.add_trace(go.Bar(
    x=no_laborales_por_mes['Mes'],
    y=no_laborales_por_mes['Cantidad Pedida'],
    name='No Laborales',
    text=no_laborales_por_mes['Cantidad Pedida'],
    marker_color='red'
))

# Línea en el medio.
linea_media = sum(dias_laborales_por_mes['Cantidad Pedida']) / len(dias_laborales_por_mes)
fig.add_shape(
    type="line",
    x0=min(dias_laborales_por_mes['Mes']),
    x1=max(dias_laborales_por_mes['Mes']),
    y0=linea_media,
    y1=linea_media,
    line=dict(color="black", width=2, dash="dash")
)

# Personalizar el diseño del gráfico.
fig.update_layout(
    xaxis_title='Mes',
    yaxis_title='Cantidad Vendida',
    title='Cantidades pedidas por Mes en Días Laborales y No Laborales',
    xaxis=dict(tickmode='array', tickvals=list(range(1, 13)),
               ticktext=meses)
)

# Mostrar el gráfico.
fig.show()

6. Impacto de eventos especiales en las ventas:
● ¿Se ha observado algún aumento o disminución significativa en las ventas en días
cercanos a eventos especiales, como días festivos?
● ¿Qué eventos específicos han tenido un impacto notable en el comportamiento de
las ventas y cómo se manifestó ese impacto?

Agrupado por dias y sumadas las ventas en dias iguales

Solo un campo con los dias

In [240]:
df_por_dias = dataframe_total # realizo una copia el df con toda la informacion del año para no modicfiocar el original

# como la fecha esta en un string la convierte en Data Type
df_por_dias['Fecha de Pedido'] = pd.to_datetime(df_por_dias['Fecha de Pedido'])

df_por_dias['solo_dia']=df_por_dias['Fecha de Pedido'].dt.date

df_por_dias

ID de Pedido                    Producto  Cantidad Pedida  \
0            141234                      iPhone                1   
1            141235    Lightning Charging Cable                1   
2            141236            Wired Headphones                2   
3            141237            27in FHD Monitor                1   
4            141238            Wired Headphones                1   
...             ...                         ...              ...   
185945       319666    Lightning Charging Cable                1   
185946       319667       AA Batteries (4-pack)                2   
185947       319668             Vareebadd Phone                1   
185948       319669            Wired Headphones                1   
185949       319670  Bose SoundSport Headphones                1   

        Precio Unitario     Fecha de Pedido  \
0                700.00 2019-01-22 21:25:00   
1                 14.95 2019-01-28 14:15:00   
2                 11.99 2019-01-17 13:33:00   
3                149.99 2019-01-05 20:33:00   
4                 11.99 2019-01-25 11:59:00   
...                 ...                 ...   
185945            14.95 2019-12-11 20:58:00   
185946             3.84 2019-12-01 12:01:00   
185947           400.00 2019-12-09 06:43:00   
185948            11.99 2019-12-03 10:39:00   
185949            99.99 2019-12-21 21:45:00   

                            Dirección de Envio  Total venta  Hora  Dia  \
0              944 Walnut St, Boston, MA 02215       700.00    21   22   
1             185 Maple St, Portland, OR 97035        14.95    14   28   
2        538 Adams St, San Francisco, CA 94016        23.98    13   17   
3           738 10th St, Los Angeles, CA 90001       149.99    20    5   
4                387 10th St, Austin, TX 73301        11.99    11   25   
...                                        ...          ...   ...  ...   
185945  14 Madison St, San Francisco, CA 94016        14.95    20   11   
185946    549 Willow St, Los Angeles, CA 90001         7.68    12    1   
185947        273 Wilson St, Seattle, WA 98101       400.00     6    9   
185948          778 River St, Dallas, TX 75001        11.99    10    3   
185949  747 Chestnut St, Los Angeles, CA 90001        99.99    21   21   

          solo_dia  Anio  Mes Dia_semana  
0       2019-01-22  2019    1    Tuesday  
1       2019-01-28  2019    1     Monday  
2       2019-01-17  2019    1   Thursday  
3       2019-01-05  2019    1   Saturday  
4       2019-01-25  2019    1     Friday  
...            ...   ...  ...        ...  
185945  2019-12-11  2019   12  Wednesday  
185946  2019-12-01  2019   12     Sunday  
185947  2019-12-09  2019   12     Monday  
185948  2019-12-03  2019   12    Tuesday  
185949  2019-12-21  2019   12   Saturday  

[185916 rows x 13 columns]

In [241]:
# Agrupo por dias y suma todos las cantidades de  ventas en ese dia
df_suma_diaria = df_por_dias.groupby('solo_dia')['Cantidad Pedida'].sum().reset_index() 

# Genera otro df con los totales precio de ventas por dias
df_ventas_totales_dia = df_por_dias.groupby('solo_dia')['Total venta'].sum().reset_index() 

df_suma_diaria,df_ventas_totales_dia

(       solo_dia  Cantidad Pedida
 0    2019-01-01              343
 1    2019-01-02              368
 2    2019-01-03              330
 3    2019-01-04              330
 4    2019-01-05              355
 ..          ...              ...
 360  2019-12-27              869
 361  2019-12-28              931
 362  2019-12-29              954
 363  2019-12-30              928
 364  2019-12-31              886
 
 [365 rows x 2 columns],
        solo_dia  Total venta
 0    2019-01-01     65681.94
 1    2019-01-02     70813.20
 2    2019-01-03     47046.20
 3    2019-01-04     62012.21
 4    2019-01-05     46524.63
 ..          ...          ...
 360  2019-12-27    126628.05
 361  2019-12-28    134015.50
 362  2019-12-29    156024.62
 363  2019-12-30    152319.81
 364  2019-12-31    131454.30
 
 [365 rows x 2 columns])

Dias festivos en los EEUU

In [242]:
festivos_federales_EEUU = pd.DataFrame({
    'Conmemora': ['Año Nuevo', 'Día de Martin Luther King Jr.', 'Día de los Presidentes', 'Día de la Recordación', 'Día de la Independencia', 'Día del Trabajo', 'Día de Colón', 'Día de Acción de Gracias', 'Navidad'],
    'Fecha': ['2019-01-01', '2019-01-16', '2019-02-20', '2019-05-29', '2019-07-04', '2019-09-04', '2019-10-09', '2019-11-23', '2019-12-25']
})
dias_familiares_EEUU = pd.DataFrame({
    'Conmemora': ['Día de la Madre', 'Día del Padre', 'Día del Niño', 'Día del Abuelo'],
    'Fecha': ['2019-05-14', '2019-06-18', '2019-06-11', '2019-09-10']
})


festivos_totales = pd.concat([festivos_federales_EEUU, dias_familiares_EEUU], ignore_index=True)
festivos_totales

Conmemora       Fecha
0                       Año Nuevo  2019-01-01
1   Día de Martin Luther King Jr.  2019-01-16
2          Día de los Presidentes  2019-02-20
3           Día de la Recordación  2019-05-29
4         Día de la Independencia  2019-07-04
5                 Día del Trabajo  2019-09-04
6                    Día de Colón  2019-10-09
7        Día de Acción de Gracias  2019-11-23
8                         Navidad  2019-12-25
9                 Día de la Madre  2019-05-14
10                  Día del Padre  2019-06-18
11                   Día del Niño  2019-06-11
12                 Día del Abuelo  2019-09-10

In [243]:
# DF DE DiAS FESTIVOS EN LOS EEUU

# Separar el número de día y el mes
df_dias_festivos=festivos_totales

# Convierto a tipo date el campo fecha
df_dias_festivos['Fecha'] = pd.to_datetime(df_dias_festivos['Fecha'])
df_dias_festivos['solo_dia']=df_dias_festivos['Fecha'].dt.date

# Crea 2 columnas de meses y dias respectivamente 
df_dias_festivos['Dia'] = df_dias_festivos['solo_dia'].apply(lambda x: x.day)
df_dias_festivos['Mes'] = df_dias_festivos['solo_dia'].apply(lambda x: x.month)

# Ordena el df por dias y luego por meses, de esa manera queda cronologicamente ordenado
df_dias_festivos = df_dias_festivos.sort_values(by='Dia')
df_dias_festivos = df_dias_festivos.sort_values(by='Mes')


# de aca se genran las listas de dias separadas por meses para el grafico de CALOR

# Iterar por cada mes único
meses_unicos = df_dias_festivos['Mes'].unique()  # genere un array con los numeros de los meses
niveles_festivos =[]

for mes in meses_unicos:
    # Filtrar el DataFrame por mes
    df_mes = df_dias_festivos[df_dias_festivos['Mes'] == mes]
    
    # Obtener los valores de niveles para cada día en una lista
    niveles_festivos.append(df_mes['Conmemora'].tolist())
    

df_dias_festivos,niveles_festivos

(                        Conmemora      Fecha    solo_dia  Dia  Mes
 0                       Año Nuevo 2019-01-01  2019-01-01    1    1
 1   Día de Martin Luther King Jr. 2019-01-16  2019-01-16   16    1
 2          Día de los Presidentes 2019-02-20  2019-02-20   20    2
 9                 Día de la Madre 2019-05-14  2019-05-14   14    5
 3           Día de la Recordación 2019-05-29  2019-05-29   29    5
 11                   Día del Niño 2019-06-11  2019-06-11   11    6
 10                  Día del Padre 2019-06-18  2019-06-18   18    6
 4         Día de la Independencia 2019-07-04  2019-07-04    4    7
 5                 Día del Trabajo 2019-09-04  2019-09-04    4    9
 12                 Día del Abuelo 2019-09-10  2019-09-10   10    9
 6                    Día de Colón 2019-10-09  2019-10-09    9   10
 7        Día de Acción de Gracias 2019-11-23  2019-11-23   23   11
 8                         Navidad 2019-12-25  2019-12-25   25   12,
 [['Año Nuevo', 'Día de Martin Luther King Jr.'

Grafico de ventas por dias

In [244]:
fig = go.Figure()

altura_maxima = int(df_suma_diaria['Cantidad Pedida'].max())
# Grafico de barras
fig.add_trace(go.Bar(
        x= df_suma_diaria['solo_dia'],
        y= df_suma_diaria['Cantidad Pedida'],
        yaxis="y",
       
        marker = dict(color= df_suma_diaria['Cantidad Pedida'], colorscale = 'RdYlGn'),
        text = '',
        textposition = 'auto',
        showlegend=False
))

# Marcas de triangulos de los dias festivos


for index, row in festivos_totales.iterrows():
    fecha = row['Fecha']
    conmemoracion = row['Conmemora']
    descripcion = f' {fecha}<br> {conmemoracion}'  # Descripción con fecha y valor
    fig.add_trace(
        go.Scatter(
            x=[fecha],
            y=[df_suma_diaria['Cantidad Pedida'].max()],  # Altura del triangulo correspondiente a os valores del otro grafico aprox 1000
            mode='markers',
            marker=dict(symbol='triangle-down', size=10, color='red',line=dict(width=1, color='black')),  # Triangulo rojo invertido Borde personalizado
            showlegend=False 
            
            # name=descripcion
        )
    )
    fig.add_annotation(
        go.layout.Annotation(
            x=fecha,
            y=df_suma_diaria['Cantidad Pedida'].max(),
            text=descripcion,
            showarrow=False,
            font=dict(color='blue', size=6),
            xanchor='center',  # Centrar la etiqueta en la fecha
            yanchor='bottom',  # Colocar la etiqueta sobre el triángulo
            textangle=280,      # Orientación de 45 grados
            
        )
    )



fig.update_layout(
    title = 'Cantidad de productos vendidos por dia',
    # xaxis = dict(title='Dias'),
    yaxis = dict(title='Cantidad', side='left'),
    xaxis_title='Meses',
    # rango del eje y con un 10% mas del min y max valor
    yaxis_range = [0, df_suma_diaria['Cantidad Pedida'].max() * 1.1],
    xaxis=dict(
        range=['2018-12-28', '2020-01-08'], # Establecer el rango de fechas en el eje X
        
        # tickvals=list(range(12)),  # Ubicaciones de las etiquetas de los meses (0 a 11)
        ticktext=meses,           # Nombres de los meses
        tickangle=35,
    )
    
)
# Muestro el grafico
fig.show()

SE SEPARAN LOS DF MARCANDO LOS MESES U DIAS POR SEPARADOS PARA GENERAR LISTAS QUE SE UTILIZARAN EN EL GRAFICO DE CALOR

In [245]:
# DF DE CANTIAD e productos vendidos POR DIA

# Separar el número de día y el mes
df_dia_mes_valor=df_suma_diaria

# genera el campo con el numero de dia
df_dia_mes_valor['Dia'] = df_dia_mes_valor['solo_dia'].apply(lambda x: x.day)
# genera el campo con el numero de mes
df_dia_mes_valor['Mes'] = df_dia_mes_valor['solo_dia'].apply(lambda x: x.month)

df_dia_mes_valor.drop(df_dia_mes_valor[(df_dia_mes_valor['solo_dia']=="2019-01-01")].index,inplace=True) # se borra el primer dia del año 2020
# Elimina el ultimo registro que corresponde al 01 de enero del 2020, para evitar errore en el grafico
# if df_dia_mes_valor(index=365) in df_dia_mes_valor:
#     df_dia_mes_valor = df_dia_mes_valor.drop(index=365)

# Iterar por cada mes único
meses_unicos = df_dia_mes_valor['Mes'].unique()  # genere un array con los numeros de los meses
niveles_por_dia =[]

for mes in meses_unicos:
    # Filtrar el DataFrame por mes
    df_mes = df_dia_mes_valor[df_dia_mes_valor['Mes'] == mes]
    
    # Obtener los valores de niveles para cada día en una lista
    niveles_por_dia.append(df_mes['Cantidad Pedida'].tolist())

df_dia_mes_valor

solo_dia  Cantidad Pedida  Dia  Mes
0    2019-01-01              343    1    1
1    2019-01-02              368    2    1
2    2019-01-03              330    3    1
3    2019-01-04              330    4    1
4    2019-01-05              355    5    1
..          ...              ...  ...  ...
360  2019-12-27              869   27   12
361  2019-12-28              931   28   12
362  2019-12-29              954   29   12
363  2019-12-30              928   30   12
364  2019-12-31              886   31   12

[365 rows x 4 columns]

In [246]:
# Iterar por cada mes único
meses_unicos = df_dia_mes_valor['Mes'].unique()  # genere un array con los numeros de los meses
niveles_por_dia =[]

for mes in meses_unicos:
    # Filtrar el DataFrame por mes
    df_mes = df_dia_mes_valor[df_dia_mes_valor['Mes'] == mes]
    
    # Obtener los valores de niveles para cada día en una lista
    niveles_por_dia.append(df_mes['Cantidad Pedida'].tolist())

df_dia_mes_valor

solo_dia  Cantidad Pedida  Dia  Mes
0    2019-01-01              343    1    1
1    2019-01-02              368    2    1
2    2019-01-03              330    3    1
3    2019-01-04              330    4    1
4    2019-01-05              355    5    1
..          ...              ...  ...  ...
360  2019-12-27              869   27   12
361  2019-12-28              931   28   12
362  2019-12-29              954   29   12
363  2019-12-30              928   30   12
364  2019-12-31              886   31   12

[365 rows x 4 columns]

In [247]:
#  Mapa de calor, de la cantida de ventas realizadas por dia
fig = go.Figure(data=go.Heatmap(z=niveles_por_dia))

# Personalizar el diseño del mapa de calor
fig.update_layout(
    title='Ventas Totales cada dia',
    xaxis_title='Dias',
    yaxis_title='Meses'
)


# Agregar las marcas de los dias festivos
for index, row in df_dias_festivos .iterrows():
    x = row['Dia']
    y = row['Mes']
    texto = row['Conmemora']
    
    fig.add_annotation(
        # a las coordenadas de ias y meses se le resta 1 porque el grafico comienza con 0
        x=x-1,
        y=y-1,
        text=texto,
        showarrow=True,
        arrowhead=2,
        ax=20, # desplazamiento en eje x
        ay=-20, # desplazamiento en eje y
        font=dict(color='white'),  #  color del texto
        bgcolor='green',  #  fondo de la marca
        opacity=0.4  #  opacidad de la marca
    )

# Texto de los meses en eje Y
fig.update_yaxes(
    tickvals=list(range(len(meses))),  # Ubicaciones de las etiquetas en el eje Y
    ticktext=meses,  # Texto de las etiquetas en el eje Y
)

# Texto de los dias en eje X
fig.update_xaxes(
    tickvals=list(range(len(dias))),  # Ubicaciones de las etiquetas en el eje X
    ticktext=dias,  # Texto de las etiquetas en el eje X
)

# Mostrar el mapa de calor
fig.show()


Grafico de calor con valores totales vendidos por dias

In [248]:
# Separar el número de día y el mes
df_dia_mes_valor=df_suma_diaria

# genera el campo con el numero de dia
df_ventas_totales_dia['Dia'] = df_ventas_totales_dia['solo_dia'].apply(lambda x: x.day)
# genera el campo con el numero de mes
df_ventas_totales_dia['Mes'] = df_ventas_totales_dia['solo_dia'].apply(lambda x: x.month)

# Elimina el ultimo registro que corresponde al 01 de enero del 2020, para evitar errore en el grafico
df_dia_mes_valor.drop(df_dia_mes_valor[(df_dia_mes_valor['solo_dia']=="2019-01-01")].index,inplace=True) # se borra el primer dia del año 2020
#df_ventas_totales_dia = df_ventas_totales_dia.drop(index=365)
# Iterar por cada mes único
meses_unicos = df_ventas_totales_dia['Mes'].unique()  # genere un array con los numeros de los meses
totales_ventas_por_dia =[]

for mes in meses_unicos:
    # Filtrar el DataFrame por mes
    df_mes = df_ventas_totales_dia[df_ventas_totales_dia['Mes'] == mes]
    
    # Obtener los valores de niveles para cada día en una lista
    totales_ventas_por_dia.append(df_mes['Total venta'].tolist())

totales_ventas_por_dia

[[65681.94,
  70813.2,
  47046.2,
  62012.21,
  46524.63,
  52777.49,
  53676.42,
  56112.47,
  55153.13,
  56660.92,
  78414.54,
  48126.41,
  61284.39,
  50090.48,
  64869.32,
  51305.0,
  55152.75,
  48469.74,
  56964.73,
  68084.88,
  60948.29,
  59380.92,
  57572.45,
  55848.86,
  57721.49,
  71476.18,
  63683.97,
  57419.48,
  61692.21,
  56238.61,
  62383.13],
 [71757.33,
  77893.74,
  72971.05,
  70260.0,
  76124.54,
  86290.9,
  74316.68000000001,
  94507.13,
  87770.12,
  76934.51,
  75116.02,
  81070.51,
  92570.14,
  71620.69,
  75581.65,
  87786.35,
  84481.74,
  80362.52,
  75558.98,
  75050.82,
  80969.31,
  82520.99,
  65925.8,
  74934.91,
  71061.15,
  80006.81,
  90377.62,
  68200.41],
 [90943.91,
  92309.7,
  85518.78,
  88485.21,
  100209.44,
  86296.07,
  95664.75,
  91815.88,
  94645.37,
  83418.55,
  88651.19,
  90488.99,
  88235.73,
  82178.02,
  84574.49,
  81095.76,
  81371.81,
  100970.47,
  85103.11,
  86559.98,
  85403.07,
  102992.11,
  89978.4,
  78756.76

Grafico de calor en de cada dia del año con la sumatoria de los montos de las ventas realizadas y las marcas de los dias festivos

In [249]:
#  Mapa de calor, de la cantida de ventas realizadas por dia
fig = go.Figure(data=go.Heatmap(z=totales_ventas_por_dia))

# Personalizar el diseño del mapa de calor
fig.update_layout(
    title='Montos totales de ventas por dia',
    xaxis_title='Dias',
    yaxis_title='Meses'
)


# Agregar las marcas de los dias festivos
for index, row in df_dias_festivos .iterrows():
    x = row['Dia']
    y = row['Mes']
    texto = row['Conmemora']
    
    fig.add_annotation(
        # a las coordenadas de ias y meses se le resta 1 porque el grafico comienza con 0
        x=x-1,
        y=y-1,
        text=texto,
        showarrow=True,
        arrowhead=2,
        ax=20, # desplazamiento en eje x
        ay=-20, # desplazamiento en eje y
        font=dict(color='white'),  #  color del texto
        bgcolor='green',  #  fondo de la marca
        opacity=0.4  #  opacidad de la marca
    )

# Texto de los meses en eje Y
fig.update_yaxes(
    tickvals=list(range(len(meses))),  # Ubicaciones de las etiquetas en el eje Y
    ticktext=meses,  # Texto de las etiquetas en el eje Y
)

# Texto de los dias en eje X
fig.update_xaxes(
    tickvals=list(range(len(dias))),  # Ubicaciones de las etiquetas en el eje X
    ticktext=dias,  # Texto de las etiquetas en el eje X
)

# Mostrar el mapa de calor
fig.show()


PREGUNTA ADICIONAL

Análisis de productos cruzados: Analizar qué productos tienden a venderse juntos, lo que podría llevar a oportunidades de ventas adicionales o promociones.

In [257]:

from itertools import combinations

# Utilizando el df con las ventas duarias de todo el año se buscan los valores de ventas diarias como referencias cruzadas


# Crear una tabla de pivote para obtener la cantidad de cada producto vendido por día
tabla_pivote = df_por_dias.pivot_table(index='solo_dia', columns='Producto', values='Cantidad Pedida', aggfunc='sum', fill_value=0)

# Se analizan todas las combinacione posibles de productos, y se toman los grupos de ventas por dias tomadas en conjunto.

# Calcular el lift entre los productos
lifts = {}  # Diccionario para almacenar los lifts
productos = df_por_dias['Producto'].unique()  # Lista de productos únicos

for par in combinations(productos, 2):  # Combinaciones de 2 productos a la vez
    producto_A, producto_B = par

    # Calcular el soporte conjunto
    soporte_conjunto = tabla_pivote[producto_A].mul(tabla_pivote[producto_B]).sum()

    # Calcular el soporte individual de los productos
    soporte_Producto_A = tabla_pivote[producto_A].sum()
    soporte_Producto_B = tabla_pivote[producto_B].sum()

    # Calcular el lift y guardarlo en el diccionario
    lift = (soporte_conjunto / len(df)) / (soporte_Producto_A / len(df)) / (soporte_Producto_B / len(df))
    lifts[par] = lift

# Encontrar el par con el lift más alto
par_con_lift_mas_alto = max(lifts, key=lifts.get)
lift_mas_alto = lifts[par_con_lift_mas_alto]

# Encontrar el par mas bajo
par_con_lift_mas_bajo = min(lifts, key=lifts.get)
lift_mas_bajo = lifts[par_con_lift_mas_bajo]

print("El par con el lift más alto es:", par_con_lift_mas_alto)
print("El valor del lift más alto es:", lift_mas_alto)
print("El par con el lift más bajo es:", par_con_lift_mas_bajo)
print("El valor del lift más bajo es:", lift_mas_bajo)

El par con el lift más alto es: ('Vareebadd Phone', 'LG Dryer')
El valor del lift más alto es: 34.63051400378347
El par con el lift más bajo es: ('LG Dryer', 'LG Washing Machine')
El valor del lift más bajo es: 32.43123309067582


Valores para graficar

In [260]:
productosLista=productos.tolist()
productosLista

['iPhone',
 'Lightning Charging Cable',
 'Wired Headphones',
 '27in FHD Monitor',
 'AAA Batteries (4-pack)',
 '27in 4K Gaming Monitor',
 'USB-C Charging Cable',
 'Bose SoundSport Headphones',
 'Apple Airpods Headphones',
 'Macbook Pro Laptop',
 'Flatscreen TV',
 'Vareebadd Phone',
 'AA Batteries (4-pack)',
 'Google Phone',
 '20in Monitor',
 '34in Ultrawide Monitor',
 'ThinkPad Laptop',
 'LG Dryer',
 'LG Washing Machine']

In [277]:
import plotly.express as px


# Crear un DataFrame a partir del diccionario de lifts
df_lifts = pd.DataFrame(list(lifts.items()), columns=['Par de Productos', 'Lift'])
df_lifts[['Producto_A', 'Producto_B']] = pd.DataFrame(df_lifts['Par de Productos'].tolist(), index=df_lifts.index)

# Crear una tabla de pivote para tener los productos en los ejes x e y
pivot_df = df_lifts.pivot(index='Producto_A', columns='Producto_B', values='Lift')

# Crear el mapa de calor con Plotly
fig = px.imshow(pivot_df)
fig.update_xaxes(side="top")  # Mover los nombres de los productos al lado superior del eje x
fig.update_layout(
    title="Mapa de Calor de Valores de Lift",
    title_x=0.2 , # Ajustar la posición horizontal del título (0.5 significa centrado)
    title_y=0.95,  # Ajustar la posición vertical del título (0.95 significa cerca de la parte superior)
    title_xanchor="center",  # Anclaje horizontal del título (centro)
    title_yanchor="top", # Anclaje vertical del título (parte superior)
    xaxis_title="Productos",
    yaxis_title="Productos",
    width=800,  # Ancho en píxeles
    height=600  # Alto en píxeles
)

fig.show()

19